### ISSUE TITLE
WebGL's bufferData and bufferSubData should accept ArrayBufferView

### ISSUE THREAD

**Issue by dmikis** [microsoft/TypeScript#24195]
**TypeScript Version:**  2.8.3

**Search Terms:** BufferSource

**Code**

```ts
function foo(gl: WebGLRenderingContext, data: ArrayBuffer | ArrayBufferView): void {
    gl.bufferSubData(gl.ARRAY_BUFFER, 0, data);
}
```

**Expected behavior:**

No errors

**Actual behavior:**

```
Argument of type 'ArrayBuffer | ArrayBufferView' is not assignable to parameter of type 'Float32Array | Int8Array | Int16Array | Int32Array | Uint8Array | Uint16Array | Uint32Array | Uin...'.
  Type 'ArrayBufferView' is not assignable to type 'Float32Array | Int8Array | Int16Array | Int32Array | Uint8Array | Uint16Array | Uint32Array | Uin...'.
    Type 'ArrayBufferView' is not assignable to type 'ArrayBuffer'.
      Property 'slice' is missing in type 'ArrayBufferView'.
```

**Playground Link:** [link](http://www.typescriptlang.org/play/index.html#src=function%20foo(gl%3A%20WebGLRenderingContext%2C%20data%3A%20ArrayBuffer%20%7C%20ArrayBufferView)%3A%20void%20%7B%0A%20%20%20%20gl.bufferSubData(gl.ARRAY_BUFFER%2C%200%2C%20data)%3B%0A%7D)

**Related Issues:** -

I believe `ArrayBufferView` interface should be removed in favour of an alias to union of typed array types and `DataView`, as it's [defined](https://heycam.github.io/webidl/#ref-for-ArrayBufferView) in the Web platform.

**Comment by mhegazy** [comment]
The issue is how ArrayBufferView is defined in the library file. the spec specifies `ArrayBufferView` as a union type, see [definition](https://heycam.github.io/webidl/#ArrayBufferView). The function definition is correct, but `ArrayBufferView` is defined as an interface and not as a union type.

My recommendation is not not use `ArrayBufferView` in your explicit type annotation, and instead use a union of possible typed arrays.

For the long term, we need to update the definition of ArrayBufferView, though this will be a breaking change.

**Comment by dmikis** [comment]
Could `BufferSource` be not defined via current `ArrayBufferView`?

**Comment by jakebailey** [comment]
Just looking at old issues; this was fixed in #26102.
